### DISCUSSION TITLE
[RFC] New Flex component

### DISCUSSION THREAD

**Discussion by zombieJ** [ant-design/ant-design#44070] [RFCs - archive]
# 📋 RFC

## Summary

Create new component Flex for the flex layout to simplify Row & Col usage.

## Basic example

```tsx
import { Flex } from 'antd';

<Flex gutter={[32, 'small']}>
  <div />
  <div />
</Flex>
```

## Motivation

Row & Col component is heavy in some case when we only need simplify flex view which we should provide a simple way like Space component.

It's easy to apply nest case:

```tsx
<Flex gap={16} align="stretch">
  <Avatar />
  <Flex vertical>
    <h3>Username</h3>
    <p style={{ flex: 'auto' }}>Description</p>
  </Flex>
</Flex>
```

And for Grid:

* Row `align` not support `stretch` since children must be `Col`.
* Row is semantically means **ROW**, it should not be **Col** for above case.

A fallback case for hard writing:

```tsx
<Row style={{ align: 'stretch'  }} gutter={16}>
  <Col> // Must use `Col` since `gutter` only work on `Col`
    <Avatar />
  </Col>
  <Col flex="auto">
    <Row gutter={[16, 16]}> // Same need use `Col` since `gutter` not support
      <Col span={24}>
        <h3>Username</h3>
      </Col>
      <Col span={24}>  // Here can not apply `flex: auto` since it's not column flex
        <p style={{ flex: 'auto' }}>Description</p>
      </Col>
    </Row>
  </Col>
</Row>
```

## Detailed

Flex behavior:
* Support direction. Default: Horizontal
* Support break line. Default: no break
* Support align. Default: center
* Support justify. Default: start align

## Drawbacks

New component.

## Adoption strategy

Apply new component in minor version.

**Comment by crazyair** [comment]
如果是这样的代码，gutter 配置就比较难受了
```tsx
<Grid gutter={[32, 'small']}>
  {isShow && <div />}
  <div />
</Grid>
```

**Reply by crazyair** [reply]
一切配置都可以是 `items`，类似 Menu...
```tsx
<Grid items={[{ gutter:32,  render: <div/>, isShow }, { gutter:'small', render: <div/> }]} />
```

**Reply by zombieJ** [reply]
Grid 本身其实就是 flex 布局，没做什么特别的事情。主要是平替 Row & Col，这两个组件用起来太复杂了

**Reply by afc163** [reply]
别 items，items 方便直接消费远程数据。布局都是手写的，用 items 反而不直观。

**Reply by crazyair** [reply]
gutter API 我理解错了，以为是 children 每一项的 size

**Comment by li-jia-nan** [comment]
- #44362

**Comment by manivsh** [comment]
also, guys, why does none of the antd grid/layout have anything for cols elements that have different heights (like css grid-row: span x), for making grids like 👇 
![image](https://github.com/ant-design/ant-design/assets/134940591/8d9ad1c3-7eb2-4584-be19-d8d24e44c824)
  
`<Space />`, align elements relative to rows (which height is max height of elements)

current `<Grid />`, require `<Row />` which aligns content relative to itself.

**Reply by zombieJ** [reply]
Row is horizontal level layout. It can not apply this layout. But Flex can do like:

```tsx
<Flex align="stretch">
  <Flex vertical>
    <div style={{ height1 }} />
    <div style={{ height2 }} />
  </Flex>
  <Flex vertical>
    <div style={{ height3 }} />
    <div style={{ flex: 'auto' }} />
  </Flex>
</Flex>
```
